autrainer/training/utils.py

```diff
@@ -0,0 +1,100 @@
+import os
+from typing import Dict, Optional
+
+import torch
+
+
+def format_results(
+    results: Dict[str, float],
+    results_type: str,
+    training_type: str,
+    iteration: Optional[int] = None,
+) -> str:
+    s = f"{results_type} results"
+    s += f" at {training_type} {iteration}" if iteration else ""
+    s += ":\n"
+    max_key_len = max([len(k) for k in results.keys()])
+    s += "\n".join(
+        [f"{(k+':').ljust(max_key_len+1)} {v:.4f}" for k, v in results.items()]
+    )
+    return s
+
+
+def load_pretrained_model_state(
+    model: torch.nn.Module,
+    state_dict: Dict[str, torch.Tensor],
+    skip_last_layer: bool = False,
+) -> bool:
+    """Load a pretrained model state dict for a model and skip the last linear
+    layer if the output dimension of the model differs or skip_last_layer is
+    True.
+
+    Args:
+        model: Model to load the state dict into.
+        state_dict: State dict to load into the model.
+        skip_last_layer: Whether to always skip the last linear layer when
+            loading the state dict, irrespective of the shape.
+            Defaults to False.
+
+    Returns:
+        True if the last linear layer was skipped, otherwise False.
+    """
+    last_linear_layer = None
+    for name, layer in model.named_modules():
+        if isinstance(layer, torch.nn.Linear):
+            last_linear_layer = name
+
+    if last_linear_layer is None:
```

Comment:
I don't think we should raise an error if there is no linear layer? Could be a fully convolutinal NN with a final pooling layer and sigmoid for binary classification for instance that should be a viable option to load too?

Reply:
Good point. In that case loading would fail. Instead of throwing an error we could wrap `model.load_state_dict` in a try catch block and fail if we are unable to load the features.
However if there is no linear layer present, it is hard to find out what has to be removed from the state dict.
In theory we can also traverse the `named_modules` of the model along the weights and biases of the state dict and check for matching shapes, then keep only the weight up until the last matching shape.
What approach would you perfer?

Reply:
Sounds good, then i will change it to a try catch block for now, and later on find a more generalized solution (for e.g. fully convolutional networks) once we have a concrete use case.